h2o-algos/src/main/java/hex/tree/SharedTreeModel.java

```diff
@@ -736,5 +738,110 @@ public void generate(JCodeSB out) {
+    // remove non-feature columns
+    adaptFrm.remove(_parms._response_column);
+    adaptFrm.remove(_parms._fold_column);
+    adaptFrm.remove(_parms._weights_column);
+    adaptFrm.remove(_parms._offset_column);
+
+    final String[] outputNames = ArrayUtils.append(adaptFrm.names(), "BiasTerm");
+    return new ScoreContributionsTask(this, _output._ntrees, _output._treeKeys, _output._init_f)
+            .doAll(outputNames.length, Vec.T_NUM, adaptFrm)
+            .outputFrame(destination_key, outputNames, null);
+  }
+
+
+  public class ScoreContributionsTask extends MRTask<ScoreContributionsTask> {
+
+    private final Key<SharedTreeModel> _modelKey;
+
+    private transient SharedTreeModel _model;
+    private transient int _ntrees;
+    private transient Key<CompressedTree>[/*_ntrees*/][/*_nclass*/] _treeKeys;
+    private transient double _init_f;
+    private transient TreeSHAPPredictor<double[]> _treeSHAP;
+
+    public ScoreContributionsTask(SharedTreeModel model, int ntrees,
+                                  Key<CompressedTree>[/*_ntrees*/][/*_nclass*/] treeKeys,
+                                  double init_f) {
+      _modelKey = model._key;
+      _ntrees = ntrees;
+      _treeKeys = treeKeys;
+      _init_f = init_f;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void setupLocal() {
+      _model = _modelKey.get();
+      assert _model != null;
+      final SharedTreeNode[] empty = new SharedTreeNode[0];
+      List<TreeSHAPPredictor<double[]>> treeSHAPs = new ArrayList<>(_ntrees);
+      for (int treeIdx = 0; treeIdx < _ntrees; treeIdx++) {
+        for (int treeClass = 0; treeClass < _treeKeys[treeIdx].length; treeClass++) {
+          if (_treeKeys[treeIdx][treeClass] == null) {
+            continue;
+          }
+          SharedTreeSubgraph tree = _model.getSharedTreeSubgraph(treeIdx, treeClass);
+          SharedTreeNode[] nodes = tree.nodesArray.toArray(empty);
+          treeSHAPs.add(new TreeSHAP<>(nodes, nodes, 0));
+        }
+      }
+      assert treeSHAPs.size() == _ntrees; // for now only regression and binomial to keep the output sane
+      _treeSHAP = new TreeSHAPEnsemble<>(treeSHAPs, (float) _init_f);
+    }
+
+    @Override
+    public void map(Chunk chks[], NewChunk[] nc) {
+      assert chks.length == nc.length - 1; // calculate contribution for each feature + the model bias
+      double[] input = MemoryManager.malloc8d(chks.length);
+      float[] contribs = MemoryManager.malloc4f(nc.length);
+
+      Object workspace = _treeSHAP.makeWorkspace();
+
+      for (int row = 0; row < chks[0]._len; row++) {
+        for (int i = 0; i < chks.length; i++) {
+          input[i] = chks[i].atd(row);
+        }
+        for (int i = 0; i < contribs.length; i++) {
+          contribs[i] = 0;
+        }
+
+        // calculate Shapley values
+        _treeSHAP.calculateContributions(input, contribs, 0, -1, workspace);
+
+        if (_model._parms.algoName().equals("GBM")) { // GBM
```

Comment:
I think there should be separate subclasses for this part of the method so that you can avoid doing a string comparison for every row of the frame and instatiate the correct subclass for given algo

Reply:
I have addressed this here (hopefully): https://github.com/h2oai/h2o-3/pull/3628/commits/2e5b7fa839247afd28740658bbede30382826275